symbolic-ppdb/src/format/mod.rs

```diff
@@ -359,3 +367,85 @@ pub struct Document {
     pub name: String,
     pub(crate) lang: Language,
 }
+
+/// An iterator over Embedded Sources.
+#[derive(Debug, Clone)]
+pub struct EmbeddedSourceIterator<'object, 'data> {
+    ppdb: &'object PortablePdb<'data>,
+    inner_it: CustomDebugInformationIterator<'data>,
+}
+
+impl<'object, 'data> EmbeddedSourceIterator<'object, 'data> {
+    fn new(ppdb: &'object PortablePdb<'data>) -> Result<Self, FormatError> {
+        // https://github.com/dotnet/runtime/blob/main/docs/design/specs/PortablePdb-Metadata.md#embedded-source-c-and-vb-compilers
+        let embedded_sources_kind = Uuid::parse_str("0E8A571B-6926-466E-B4AD-8AB04611F5FE")
+            .map_err(|e| FormatError::new(FormatErrorKind::InvalidStringData, e))?;
+        let inner_it = CustomDebugInformationIterator::new(ppdb, embedded_sources_kind)?;
+        Ok(EmbeddedSourceIterator { ppdb, inner_it })
+    }
+}
+
+impl<'object, 'data> Iterator for EmbeddedSourceIterator<'object, 'data> {
+    type Item = Result<EmbeddedSource<'data>, FormatError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.inner_it.next().map(|inner| {
+            inner.and_then(|info| match info.tag {
+                // Verify we got the expected tag `Document` here.
+                metadata::CustomDebugInformationTag::Document => {
+                    let document = self.ppdb.get_document(info.value as usize)?;
+                    let blob = self.ppdb.get_blob(info.blob)?;
+                    Ok(EmbeddedSource { document, blob })
+                }
+                _ => Err(FormatErrorKind::InvalidCustomDebugInformationTag(info.tag as u32).into()),
+            })
+        })
+    }
+}
+
+/// Lazy Embedded Source file reader.
+#[derive(Debug, Clone)]
+pub struct EmbeddedSource<'data> {
+    document: Document,
+    blob: &'data [u8],
+}
+
+impl<'data, 'object> EmbeddedSource<'data> {
+    /// Returns the build-time path associated with this source file.
+    pub fn get_path(&'object self) -> Cow<'object, str> {
+        Cow::Borrowed(self.document.name.as_str())
```

Comment:
```suggestion
    pub fn get_path(&'object self) -> &'object str {
        self.document.name.as_str()
```

I believe its okay to a `&str` straight away here.

Reply:
That would prevent us to change the implementation, e.g. if we later needed to return owned data.

Reply:
well indeed, though the previous problem was rather that we had to put a `Cow` inside of a derived struct because it was not allowed to return a shorter lifetime:

Parent
-> Child
    -> returning something with a parent lifetime

You usually return &str directly from functions, but might have a `Cow` internally for reasons. Though we can also keep it like it is.